File: php/src/gadgets/oauth/OAuthStore.php
<?php
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

interface OAuthStore {

	public function setOAuthServiceProviderInfo($providerKey, $providerInfo);

	public function getOAuthServiceProviderInfo($providerKey);

	public function setOAuthConsumerKeyAndSecret($providerKey, $keyAndSecret);

	public function setTokenAndSecret($tokenKey, $tokenInfo);

	public function getOAuthAccessorTokenKey(TokenKey $tokenKey);

}

class OAuthStoreVars {
	public static $HttpMethod = array('GET' => 'GET', 'POST' => 'POST');
	public static $SignatureType = array('HMAC_SHA1' => 'HMAC_SHA1', 'RSA_SHA1' => 'RSA_SHA1', 
			'PLAINTEXT' => 'PLAINTEXT');
	public static $KeyType = array('HMAC_SYMMETRIC' => 'HMAC_SYMMETRIC', 
			'RSA_PRIVATE' => 'RSA_PRIVATE');
	public static $OAuthParamLocation = array('AUTH_HEADER' => 'auth_header', 
			'POST_BODY' => 'post_body', 'URI_QUERY' => 'uri_query');
}

class AccesorInfo {
	public $accessor;
	public $httpMethod;
	public $signatureType;
	public $paramLocation;

	public function getParamLocation()
	{
		return $this->paramLocation;
	}

	public function setParamLocation($paramLocation)
	{
		$this->paramLocation = $paramLocation;
	}

	public function getAccessor()
	{
		return $this->accessor;
	}

	public function setAccessor($accessor)
	{
		$this->accessor = $accessor;
	}

	public function getHttpMethod()
	{
		return $this->httpMethod;
	}

	public function setHttpMethod($httpMethod)
	{
		$this->httpMethod = $httpMethod;
	}

	public function getSignatureType()
	{
		return $this->signatureType;
	}

	public function setSignatureType($signatureType)
	{
		$this->signatureType = $signatureType;
	}
}

class ConsumerKeyAndSecret {
	private $consumerKey;
	private $consumerSecret;
	private $keyType;

	public function ConsumerKeyAndSecret($key, $secret, $type)
	{
		$this->consumerKey = $key;
		$this->consumerSecret = $secret;
		$this->keyType = $type;
	}

	public function getConsumerKey()
	{
		return $this->consumerKey;
	}

	public function getConsumerSecret()
	{
		return $this->consumerSecret;
	}

	public function getKeyType()
	{
		return $this->keyType;
	}
}

class ProviderKey {
	private $gadgetUri;
	private $serviceName;

	public function getGadgetUri()
	{
		return $this->gadgetUri;
	}

	public function setGadgetUri($gadgetUri)
	{
		$this->gadgetUri = $gadgetUri;
	}

	public function getServiceName()
	{
		return $this->serviceName;
	}

	public function setServiceName($serviceName)
	{
		$this->serviceName = $serviceName;
	}

	public function hashCode()
	{
		$prime = 31;
		$result = 1;
		$result = $prime * $result + (($gadgetUri == null) ? 0 : $gadgetUri->hashCode());
		$result = $prime * $result + (($serviceName == null) ? 0 : $serviceName->hashCode());
		return $result;
	}

	public function equals($obj)
	{
		if ($this == $obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj . getClass())
			return false;
		$other = $obj;
		if ($gadgetUri == null) {
			if ($other->gadgetUri != null)
				return false;
		} else if (! $gadgetUri->equals(other . gadgetUri))
			return false;
		if ($serviceName == null) {
			if ($other->serviceName != null)
				return false;
		} else if (! $serviceName->equals($other->serviceName))
			return false;
		return true;
	}
}

class ProviderInfo {
	private $provider;
	private $httpMethod;
	private $signatureType;
	private $paramLocation;
	
	// this can be null if we have not negotiated a consumer key and secret
	// yet with the provider, or if we decided that we want to use a global
	// public key
	private $keyAndSecret;

	public function getParamLocation()
	{
		return $this->paramLocation;
	}

	public function setParamLocation($paramLocation)
	{
		$this->paramLocation = $paramLocation;
	}

	public function getKeyAndSecret()
	{
		return $this->keyAndSecret;
	}

	public function setKeyAndSecret($keyAndSecret)
	{
		$this->keyAndSecret = $keyAndSecret;
	}

	public function getProvider()
	{
		return $this->provider;
	}

	public function setProvider($provider)
	{
		$this->provider = $provider;
	}

	public function getHttpMethod()
	{
		return $this->httpMethod;
	}

	public function setHttpMethod($httpMethod)
	{
		$this->httpMethod = $httpMethod;
	}

	public function getSignatureType()
	{
		return $this->signatureType;
	}

	public function setSignatureType($signatureType)
	{
		$this->signatureType = $signatureType;
	}
}

class TokenKey {
	private $userId;
	private $gadgetUri;
	private $moduleId;
	private $tokenName;
	private $serviceName;

	public function getUserId()
	{
		return $this->userId;
	}

	public function setUserId($userId)
	{
		$this->userId = $userId;
	}

	public function getGadgetUri()
	{
		return $this->gadgetUri;
	}

	public function setGadgetUri($gadgetUri)
	{
		$this->gadgetUri = $gadgetUri;
	}

	public function getModuleId()
	{
		return $this->moduleId;
	}

	public function setModuleId($moduleId)
	{
		$this->moduleId = $moduleId;
	}

	public function getTokenName()
	{
		return $this->tokenName;
	}

	public function setTokenName($tokenName)
	{
		$this->tokenName = $tokenName;
	}

	public function getServiceName()
	{
		return $this->serviceName;
	}

	public function setServiceName($serviceName)
	{
		$this->serviceName = $serviceName;
	}

	public function hashCode()
	{
		$prime = 31;
		$result = 1;
		$result = $prime * $result + (($gadgetUri == null) ? 0 : $gadgetUri->hashCode());
		$result = $prime * $result + (int)($moduleId ^ ($moduleId >> 32));
		$result = $prime * $result + (($serviceName == null) ? 0 : $serviceName->hashCode());
		$result = $prime * $result + (($tokenName == null) ? 0 : $tokenName->hashCode());
		$result = $prime * $result + (($userId == null) ? 0 : $userId->hashCode());
		return $result;
	}

	public function equals($obj)
	{
		if ($this == $obj)
			return true;
		if ($obj == null)
			return false;
			/*?*/		if (getClass() != $obj->getClass())
			return false;
		
		$other = $obj;
		if ($gadgetUri == null) {
			if ($other->gadgetUri != null)
				return false;
		} else if (! $gadgetUri->equals($other->gadgetUri))
			return false;
		if ($moduleId != $other->moduleId)
			return false;
		if ($serviceName == null) {
			if ($other->serviceName != null)
				return false;
		} else if (! $serviceName->equals($other->serviceName))
			return false;
		if ($tokenName == null) {
			if ($other->tokenName != null)
				return false;
		} else if (! $tokenName->equals($other->tokenName))
			return false;
		if ($userId == null) {
			if ($other->userId != null)
				return false;
		} else if (! $userId->equals($other->userId))
			return false;
		return true;
	}
}

class TokenInfo {
	private $accessToken;
	private $tokenSecret;

	public function __construct($token, $secret)
	{
		$this->accessToken = $token;
		$this->tokenSecret = $secret;
	}

	public function getAccessToken()
	{
		return $this->accessToken;
	}

	public function getTokenSecret()
	{
		return $this->tokenSecret;
	}
}

File: java/gadgets/src/main/java/org/apache/shindig/gadgets/oauth/OAuthStore.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package org.apache.shindig.gadgets.oauth;

import org.apache.shindig.auth.SecurityToken;
import org.apache.shindig.gadgets.GadgetException;

import net.oauth.OAuthConsumer;
import net.oauth.OAuthServiceProvider;

/**
 * Interface to an OAuth Data Store. A shindig gadget server can act as an
 * OAuth consumer, using OAuth tokens to talk to OAuth service providers on
 * behalf of the gadgets it is proxying requests for. An OAuth consumer needs
 * to permanently store gadgets it has collected, and retrieve the
 * appropriate tokens when proxying a request for a gadget.
 *
 * An OAuth Data Store stores three things:
 *  (1) information about OAuth service providers, including the three
 *      URLs that define OAuth providers (defined in OAuthStore.ProviderInfo)
 *  (2) information about consumer keys and secrets that gadgets might have
 *      negotiated with OAuth service providers, or that containers might have
 *      negotiated on behalf of the gadgets. This information is bound to
 *      the service provider it pertains to and can only be stored if the
 *      corresponding service provider information is already stored in the
 *      OAuth store (defined in OAuthStore.ConsumerKeyAndSecret).
 *  (3) OAuth access tokens and their corresponding token secrets. (defined
 *      in OAuthStore.TokenInfo).
 *
 *  Note that we do not store request tokens in the OAuth store.
 */
public interface OAuthStore {

  /**
   * Information about an OAuth consumer.
   */
  public static class ConsumerInfo {
    private final OAuthConsumer consumer;
    private final String keyName;
    
    /**
     * @param consumer the OAuth consumer
     * @param keyName the name of the key to use for this consumer (passed on query parameters to
     * help with key rotation.)
     */
    public ConsumerInfo(OAuthConsumer consumer, String keyName) {
      this.consumer = consumer;
      this.keyName = keyName;
    }
    
    public OAuthConsumer getConsumer() {
      return consumer;
    }
    
    public String getKeyName() {
      return keyName;
    }
  }
  
  /**
   * Information about an access token.
   */
  public static class TokenInfo {
    private final String accessToken;
    private final String tokenSecret;
    public TokenInfo(String token, String secret) {
      accessToken = token;
      tokenSecret = secret;
    }
    public String getAccessToken() {
      return accessToken;
    }
    public String getTokenSecret() {
      return tokenSecret;
    }
  }

  /**
   * Retrieve OAuth consumer to use for requests.  The returned consumer is ready to use for signed
   * fetch requests.
   * 
   * @param securityToken token for user/gadget making request.
   * @param serviceName gadget's nickname for the service being accessed.
   * @param provider OAuth service provider info to be inserted into the returned consumer.
   * 
   * @throws GadgetException if no OAuth consumer can be found (e.g. no consumer key can be used.)
   */
  public ConsumerInfo getConsumerKeyAndSecret(SecurityToken securityToken, String serviceName,
      OAuthServiceProvider provider) throws GadgetException;

  /**
   * Retrieve OAuth access token to use for the request.
   * @param securityToken token for user/gadget making request.
   * @param consumerInfo OAuth consumer that will be used for the request.
   * @param serviceName gadget's nickname for the service being accessed.
   * @param tokenName gadget's nickname for the token to use.
   * @return the token and secret, or null if none exist
   * @throws GadgetException if an error occurs during lookup
   */
  public TokenInfo getTokenInfo(SecurityToken securityToken, ConsumerInfo consumerInfo,
      String serviceName, String tokenName) throws GadgetException;

  /**
   * Set the access token for the given user/gadget/service/token
   */
  public void setTokenInfo(SecurityToken securityToken, ConsumerInfo consumerInfo, String serviceName,
      String tokenName, TokenInfo tokenInfo) throws GadgetException;

  /**
   * Remove the access token for the given user/gadget/service/token
   */
  public void removeToken(SecurityToken securityToken, ConsumerInfo consumerInfo,
      String serviceName, String tokenName) throws GadgetException;
}
